### PR TITLE
Rendering masks

### PR DESCRIPTION
**Pull request by schell**
**tl;dr** adds `RenderLayers` to link groups of entities to cameras.

This describes which rendering layers an entity belongs to. A layer is a collection of entities that only get rendered by cameras in that layer.

Entities and cameras without a `RenderLayers` component can be considered to belong to the "default" layer (layer 0), which is the same as entities and cameras with the component `RenderLayers::default()`. 

Entities and cameras can belong to multiple layers: 

```
commands
    .spawn(...)
    .with(
        RenderLayers::layer(0) // starts with just layer 0
            .with(1) // also sets layer 1
    )
```

Using a mask of u32 we have 32 possible rendering layers, 0-31 plus the empty mask `RenderLayers::none()`.

More takeaways:
* by default, entities belong to layer 0
* entities without a mask belong to layer 0
* entities with a mask without any layers (`RenderLayers::none()`) are invisible, as if they had the component `Visible { is_visible: false }`

### CODE REVIEW

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 1]
```diff
@@ -24,22 +24,92 @@ impl VisibleEntities {
...
         let mut transparent_entities = Vec::new();
-        for (entity, visible) in visible_query.iter() {
+        for (entity, visible, maybe_ent_mask) in visible_query.iter() {
             if !visible.is_visible {
                 continue;
             }
 
+            let camera_mask = maybe_camera_mask
+                .map(|m| *m)
+                .unwrap_or_else(|| RenderingMask(0));
```
Comment (mockersf) [inline crates/bevy_render/src/camera/visible_entities.rs:105]: I think using `unwrap_or_default` would be better to define the default value only once

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 2]
```diff
@@ -24,22 +24,92 @@ impl VisibleEntities {
...
+/// A mask that describes which rendering group an entity belongs to.
+/// Cameras with this component will only render entities with a matching
+/// mask.
+#[derive(Copy, Clone, Debug, Reflect, PartialEq, Eq, PartialOrd, Ord)]
+#[reflect(Component)]
+pub struct RenderingMask(pub u32);
+
+impl Default for RenderingMask {
+    fn default() -> Self {
+        RenderingMask(0)
```
Comment (mockersf) [inline crates/bevy_render/src/camera/visible_entities.rs:36]: I think having a default mask at 1 would be better
A default at 0 means the default will match only 0, but 1 would match everything with bit 1 on, which seems better for the default camera? Not entirely sure, what do you think?

Comment (schell) [reply crates/bevy_render/src/camera/visible_entities.rs:36]: **tl;dr** Either of these change requests make sense on their own but the two together alter the logic I intended.

What the default is seems a bit arbitrary to me but I do think that "no mask" (`RenderingMask(0)`) is what entities _without_ this component should be equal to. 
The suggestion that the default be mask `1` (or group `0`) has implications about your comment above. We only want cameras with masks rendering entities with matching masks. If the camera has a mask but the entity does not, we continue the loop. If the entity has a mask but the camera does not, we continue the loop. If they both have masks but they don't match, we continue the loop.
Consider the case where we have a camera with mask `1` (group `0`) and an entity with no mask. If the default is mask `1` then the gating logic would say "these both match, since the camera defaults to mask `1`, group `0`". But what we really want is for the loop to continue since the camera has no mask at all.

If we change the default to mask `1`, group `0`, then I think the gating logic should still use `unwrap_or_else(|| RenderingMask(0))`. 

What do you think @mockersf ?

Comment (mockersf) [reply crates/bevy_render/src/camera/visible_entities.rs]: My opinions were more:

- by default, the camera should display everything, so it makes sense to be `group(0)`
- by default, all entities should be displayed, so they should be `group(0)`
- not wanting to display an entity should be an extra step, so setting its mask to `RenderingMask::default().without_group(0)`. Maybe there could be a shorter method for that

I think that makes more sense if all cameras have a mask, but if they have no mask unless the user add one yours is better

Comment (schell) [reply crates/bevy_render/src/camera/visible_entities.rs]: Ah, yes - I think that's where the confusion comes from. Being in `group(0)` means that an entity will only be rendered by cameras in `group(0)`. A camera in `group(0)` will only render entities in `group(0)`. **Without** a mask component a camera will render all entities without a mask. In a way not having a mask is a bit like being in the default group. For this reason making the default mask "empty" (`RenderingMask(0)`) makes the comparison logic very simple. 

I'm expecting folks will only use `RenderingMask` if they have a special camera that only renders some objects. For example - if you want to build a 3d user interface with a different projection than the rest of your scene. For this reason most cameras will _not_ have a mask.

**Comment by cart** [comment]
I dig this. I only have one comment: 

With the current masking approach (empty is default, empty matches empty) I don't think there is a way to encode the intent "draw everything". If you set all of the bits in the mask to 1, it will draw everything _but_ the default empty mask.  And if you set all of the bits to 0, it will only draw things that are also 0. I think it makes more sense to default to "group 1" and implicitly assume things without the mask component are also "group 1". This has the added benefit of being the behavior used in Godot, which makes it more familiar. This would mean camera masks would need to default to "group 1" (or maybe all groups, which is what godot does).

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 3]
```diff
@@ -24,22 +24,104 @@ impl VisibleEntities {
...
 
         let mut no_transform_order = 0.0;
         let mut transparent_entities = Vec::new();
-        for (entity, visible) in visible_query.iter() {
+        for (entity, visible, maybe_entity_mask) in visible_query.iter() {
             if !visible.is_visible {
                 continue;
             }
 
+            let camera_mask = maybe_camera_mask.copied().unwrap_or_default();
```
Comment (DJMcNab) [inline crates/bevy_render/src/camera/visible_entities.rs:119]: This line can probably be moved out of the inner loop.

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 4]
```diff
@@ -24,22 +24,104 @@ impl VisibleEntities {
...
+impl RenderingMask {
+    /// Create new `RenderingMask` with the given rendering group set.
+    pub fn group(n: u8) -> Self {
+        RenderingMask::default().set_group(n)
+    }
+
+    /// Set the given group on the mask.
+    /// This may be called multiple times to allow an entity to belong
+    /// to multiple rendering groups.
+    pub fn set_group(mut self, group: u8) -> Self {
```
Comment (DJMcNab) [inline crates/bevy_render/src/camera/visible_entities.rs:53]: `with_group`/`without_group` would follow the usual naming convention for these methods.

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 5]
```diff
@@ -24,22 +24,104 @@ impl VisibleEntities {
...
+/// A mask describes which rendering group an entity belongs to.
+///
+/// Cameras with this component will only render entities with a matching
+/// mask.
+///
+/// The [`Default`] instance of `RenderingMask` returns a mask that contains
+/// no groups.
+#[derive(Copy, Clone, Debug, Reflect, PartialEq, Eq, PartialOrd, Ord)]
+#[reflect(Component)]
+pub struct RenderingMask(pub u32);
```
Comment (DJMcNab) [inline crates/bevy_render/src/camera/visible_entities.rs:36]: Do we want to commit to it being a u32? We don't document what the field actually means, so it being public seems strange

Comment (schell) [reply crates/bevy_render/src/camera/visible_entities.rs:36]: It's a question of space vs max groups. I don't see there being much demand for > 32 groups, but keeping it private makes sense if we want to change it.

Comment (DJMcNab) [reply crates/bevy_render/src/camera/visible_entities.rs:36]: Yeah, keeping it private seems sensible to me, although we might want to think about how we want it to be serialised, in case we did change it. For example, changing it would break any existing scenes.

Although at this point I know that the scenes are a proof-of-concept and so the final format is likely to change anyway, so this is a bit moot presently.

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 4 above)
Comment (DJMcNab) [reply crates/bevy_render/src/camera/visible_entities.rs:53]: Also it's probably worth checking that group is valid (less than 31? 32?).

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 6]
```diff
@@ -24,22 +24,200 @@ impl VisibleEntities {
...
+    /// Panics when called with a group greater than 31.
+    pub fn without_group(mut self, group: u8) -> Self {
+        assert!(group < 32, "RenderingMask only supports groups 0 to 31");
+        self.0 |= 0 << group;
+        self
+    }
+
+    /// Get a vector of this mask's groups.
+    pub fn get_groups(&self) -> Vec<u8> {
+        (0..32)
```
Comment (cart) [inline crates/bevy_render/src/camera/visible_entities.rs:113]: The "32" here is a bit implicit (as is the u8 usage everywhere). Can we create a new `type GroupMask = u8;` type alias and use that everywhere?

Then in cases like this, can we do `std::mem::sizeof::<GroupMask> * 4`?

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 7]
```diff
@@ -24,22 +24,200 @@ impl VisibleEntities {
...
+    /// # Panics
+    /// Panics when called with a group greater than 31.
+    pub fn without_group(mut self, group: u8) -> Self {
+        assert!(group < 32, "RenderingMask only supports groups 0 to 31");
+        self.0 |= 0 << group;
+        self
+    }
+
+    /// Get a vector of this mask's groups.
+    pub fn get_groups(&self) -> Vec<u8> {
```
Comment (cart) [inline crates/bevy_render/src/camera/visible_entities.rs:112]: Can we make this return a `impl Iterator<Item=u8>` instead of `Vec<u8>`? This would cut down on unnecessary allocations and people needing a Vec can call .collect() on it.

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 8]
```diff
@@ -24,22 +24,200 @@ impl VisibleEntities {
...
+        RenderingMask(u32::MAX)
+    }
+
+    /// Create a new `RenderingMask` that belongs to no rendering groups.
+    pub fn no_groups() -> Self {
+        RenderingMask(0)
+    }
+
+    /// Create a `RenderingMask` from a list of groups.
+    pub fn from_groups(groups: &[u8]) -> Self {
```
Comment (cart) [inline crates/bevy_render/src/camera/visible_entities.rs:82]: we could make this a bit more flexible by taking a `groups: impl Iterator<Item=u8>` parameter instead

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 9]
```diff
@@ -24,22 +24,200 @@ impl VisibleEntities {
...
+    }
+
+    /// Determine if a `RenderingMask` matches another.
+    ///
+    /// `RenderingMask`s match if the first mask contains any of the groups
+    /// in the other.
+    ///
+    /// A `RenderingMask` belonging to no groups will not match any other
+    /// mask, even another belonging to no groups.
+    pub fn matches(&self, other: &RenderingMask) -> bool {
```
Comment (cart) [inline crates/bevy_render/src/camera/visible_entities.rs:125]: "Matches" is a bit ambiguous. Maybe something like `contains` or `is_superset`? Rust tends to use set terminology for collections, so my preference here is `is_superset`.

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 10]
```diff
@@ -24,22 +24,200 @@ impl VisibleEntities {
...
+}
+
+/// Defaults to a mask belonging to group `0`, the first group.
+impl Default for RenderingMask {
+    fn default() -> Self {
+        RenderingMask(1)
+    }
+}
+
+impl RenderingMask {
```
Comment (cart) [inline crates/bevy_render/src/camera/visible_entities.rs:65]: Imo "mask" is an implementation detail here. I think my preferred name is `RenderLayers` (and we could rename instances of "group" to "layer")

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 7 above)
Comment (cart) [reply crates/bevy_render/src/camera/visible_entities.rs:112]: And with the "layers" rename we could just rename this method to "iter()"

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 10 above)
Comment (cart) [reply crates/bevy_render/src/camera/visible_entities.rs:65]: In fact, this would let us simplify a bunch of method names:

* RenderMask::with_group(1) -> RenderLayers::with(1)
* RenderMask::all_groups() -> RenderLayers::all()

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 8 above)
Comment (DJMcNab) [reply crates/bevy_render/src/camera/visible_entities.rs:82]: This would be equivalent to implementing `FromIterator<u8>`, as I suggested in discord. That being said, this helper method could still be useful to have (or one using `IntoIterator`), for the likely common case of using a short-lived array, e.g. `from_groups(&[1, 3, 4])`

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 9 above)
Comment (DJMcNab) [reply crates/bevy_render/src/camera/visible_entities.rs:125]: I think the operation here is `intersects`/`intersects_with`

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 6 above)
Comment (schell) [reply crates/bevy_render/src/camera/visible_entities.rs]: I've created `type LayerMask = u32;`, `pub type Layer = u8;` and `pub const TOTAL_LAYERS = ...` to handle this.

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 9 above)
Comment (schell) [reply crates/bevy_render/src/camera/visible_entities.rs]: I changed this to `intersects`, since the function should return true if the two sets intersect.

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 10 above)
Comment (schell) [reply crates/bevy_render/src/camera/visible_entities.rs]: Done.

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 11]
```diff
@@ -24,22 +24,208 @@ impl VisibleEntities {
...
+///
+/// Cameras with this component will only render entities with intersecting
+/// layers.
+///
+/// There are 32 layers numbered `0` - [`TOTAL_LAYERS`]. Entities may belong to one or more
+/// layers, or no layer at all.
+///
+/// The [`Default`] instance of `RenderLayers` contains layer `0`, the first layer.
+///
+/// An entity with this component without any layers is invisible.
```
Comment (DJMcNab) [inline crates/bevy_render/src/camera/visible_entities.rs:42]: I'm still cautious about making this guarantee, since we still seem to require that every entity has a `Visible`, which this overlaps with

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 12]
```diff
@@ -24,22 +24,208 @@ impl VisibleEntities {
...
+/// Entities without this component belong to layer `0`.
+#[derive(Copy, Clone, Reflect, PartialEq, Eq, PartialOrd, Ord)]
+#[reflect(Component)]
+pub struct RenderLayers(LayerMask);
+
+impl std::fmt::Debug for RenderLayers {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("RenderLayers")
+            .field(&self.layers_iter().collect::<Vec<_>>())
+            .finish()
```
Comment (DJMcNab) [inline crates/bevy_render/src/camera/visible_entities.rs:53]:
```suggestion
        self.layers_iter()
            .fold(f.debug_tuple(), |t, layer| t.field(layer))
            .finish()
```
Currently, this will create a single vec field instead of a field for each layer

This suggests that this code could do with a test.

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 13]
```diff
@@ -24,22 +24,208 @@ impl VisibleEntities {
...
+    }
+}
+
+impl std::iter::FromIterator<Layer> for RenderLayers {
+    fn from_iter<T: IntoIterator<Item = Layer>>(i: T) -> Self {
+        i.into_iter().fold(RenderLayers(0), |mask, g| mask.with(g))
```
Comment (DJMcNab) [inline crates/bevy_render/src/camera/visible_entities.rs:59]: Small thought:

```suggestion
        i.into_iter().fold(Self::none(), |mask, g| mask.with(g))
```

This ensures if we change the representation, this impl keeps working.

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 14]
```diff
@@ -24,22 +24,208 @@ impl VisibleEntities {
...
+    }
+}
+
+/// Defaults to containing to layer `0`, the first layer.
+impl Default for RenderLayers {
+    fn default() -> Self {
+        RenderLayers(1)
```
Comment (DJMcNab) [inline crates/bevy_render/src/camera/visible_entities.rs:66]:
```suggestion
        RenderLayers::layer(0)
```

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 15]
```diff
@@ -24,22 +24,208 @@ impl VisibleEntities {
...
+    }
+
+    /// Create a new `RenderLayers` that belongs to no layers.
+    pub fn none() -> Self {
+        RenderLayers(0)
+    }
+
+    /// Create a `RenderLayers` from a list of layers.
+    pub fn from_layers(layers: &[Layer]) -> Self {
+        layers.iter().fold(RenderLayers(0), |mask, g| mask.with(*g))
```
Comment (DJMcNab) [inline crates/bevy_render/src/camera/visible_entities.rs:91]:
```suggestion
        layers.iter().copied().collect()
```

**Review by DJMcNab — APPROVED** [review]
Looks very good, just a few code style nits/duplications for me.

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 12 above)
Comment (schell) [reply crates/bevy_render/src/camera/visible_entities.rs:53]: Currently there is a test for the debug output. It is rendering as `"RenderLayers([0, 1, 2, 3])"`.

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 13 above)
Comment (schell) [reply crates/bevy_render/src/camera/visible_entities.rs]: Done.

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 14 above)
Comment (schell) [reply crates/bevy_render/src/camera/visible_entities.rs]: Done.

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 15 above)
Comment (schell) [reply crates/bevy_render/src/camera/visible_entities.rs]: Done.

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 12 above)
Comment (DJMcNab) [reply crates/bevy_render/src/camera/visible_entities.rs:53]: Ah, I see. I was thinking the intention was for `"RenderLayers(0, 1, 2, 3)"`, which to me seems just as reasonable. Both are fine though.

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 11 above)
Comment (schell) [reply crates/bevy_render/src/camera/visible_entities.rs:42]: Yes, it's a bit sketchy but it does help describe the feature. I think this is the best we can do at the moment. I personally plan on using the "invisiblity" of `RenderLayers::none()` so I don't think we should add a warning.

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 6 above)
Comment (cart) [reply crates/bevy_render/src/camera/visible_entities.rs]: haha good call / thanks for not just blindly plugging in `GroupMask = u8` like I asked :smile:

File: crates/bevy_render/src/camera/visible_entities.rs (hunk 11 above)
Comment (cart) [reply crates/bevy_render/src/camera/visible_entities.rs:42]: Yeah I thought a bit about just replacing Visible with RenderLayers. But I think we need a "global" visible/invisible toggle:

* That way visibility can still be toggled without needing to know what layers the current entity is visible on
* Visibility will likely eventually propagate down a hierarchy, while RenderLayers _might_ not. Its worth discussing at some point, but imo doesn't need to be settled in this pr.
* I want entity visibility to be easily toggle-able in the editor from the "Entity Inspector" (ex: have an "eye" toggle next to the entity name), which would be much harder for arbitrary layers

File: crates/bevy_render/src/camera/visible_entities.rs [hunk 16]
```diff
@@ -24,22 +24,208 @@ impl VisibleEntities {
...
+    /// # Panics
+    /// Panics when called with a layer greater than `TOTAL_LAYERS - 1`.
+    pub fn without(mut self, layer: Layer) -> Self {
+        assert!(usize::from(layer) < Self::TOTAL_LAYERS);
+        self.0 |= 0 << layer;
+        self
+    }
+
+    /// Get an iterator of the layers.
+    pub fn layers_iter(&self) -> impl Iterator<Item = Layer> {
```
Comment (cart) [inline crates/bevy_render/src/camera/visible_entities.rs:118]: In this case we're iterating a "layers collection" where imo the context is implicit. I think we should shorten this to `iter()`

**Comment by cart** [comment]
Just left one last baby comment, then I think this is good to go. Very nicely done!
